### ISSUE TITLE
[Blazor] Consider making '/src/Components/Components/src/Routing/' public

### ISSUE THREAD

**Issue by glatzert** [dotnet/aspnetcore#35067]
Currently Routing only allows one 'active' component per route. To build something which is more "angular-style"-routing, it would be convenient to have access to some of the classes from routing (especially the TemplateParsers and similar).
For my current implementation of a "nested" router, I copied all of the classes from `/src/Components/src/Routing`

So even I'm late to the party, please consider making the classes `public`.

**Comment by pranavkm** [comment]
Perhaps the issue you're describing is a dupe of #18163? If not, could you describe in detail what it is that you're trying to achieve? In general we think copying the code to produce a custom routing component is a better solution than for us to expose the implementation details of the `Router` component.

**Comment by glatzert** [comment]
I'm trying to archive angular-style routing especially in regard of multiple active components via child-routing.
e.g. /my-list will show one component with a list and have a router-outlet (or nested route view), which will be used by childroutes as e.g. /my-list/123 and show details of that specific item.

This can simply be solved by an optional Parameter, but get's complex, if you want to allow different subroutes.

/my-list
  /123
  /users/
  /groups/5
  
My plan was (and it already works as a prototype) to create a Router that can 
a) take the routing table as parameter (which allows child routes) and
b) take the current route as parameter (which allows me to remove already "used" path parts from the route:
e.g. `<NestedRouteView RouteTable="myRouteTable" CurrentRoute="myCurrentRoute" />`

Now I'll change it to use CascadingParamters, which'll make implementing NestedRouting even simpler.

Long story short: to accomplish that, I had to copy everything from the folder mentioned above and was wondering, if that's the way to go - especially for parts like detecting Route changes and parsing templates as well as converting values, which seems unneccesary to copy, since the codes already there and you'll update those bits as blazor evolves.

**Comment by javiercn** [comment]
@glatzert thanks for contacting us.

I think in your case, that's the best way to go, we want to continue adding features to the router in future versions, however, we don't in general make implementations public for the sake of convenience and recommend folks to copy the implementations if they choose to do so.

The reason is that every public bit of public API we add need to be carefully and thoughtfully considered so as to make sure that we are ok with supporting/maintaining those APIs, we don't foresee the need to change the APIs in a breaking way to add new features, and we are happy with the overall shape and structure of the APIs.

There is a cost to every API we make public, both an initial design and review cost and an ongoing maintenance cost after that, which we prefer to avoid where possible. In many implement the features in a way we don't have to expose the public API bits or recommend folks that if they can wait, produce their own implementation. In the end, the components that we ship in box are just one of the many possible implementations.

I believe this is a dupe of https://github.com/dotnet/aspnetcore/issues/27328 so I'm going to close it in favor of that one.
